### PR TITLE
RequestBodyParserMiddleware does not reject requests that exceed post_max_size

### LINKED ISSUE #254: RequestBodyBufferMiddleware should not reject requests that exceed post_max_size
If you try out example 12 (introduced via #252), you can see that the server currently *rejects* the request with a rather ugly error page if you exceed the maximum size of 100 KiB. PHP's default behavior is actually to just *ignore* the request body and still process things normally otherwise. This also allows the consumer to create a custom error message.

See http://php.net/manual/en/ini.core.php#ini.post-max-size

### PR DESCRIPTION
**Pull request by WyriHaximus**
Implements / closes #254

### CODE REVIEW

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 1]
```diff
@@ -30,27 +30,28 @@ public function __construct($sizeLimit = null)
...
             $stream->write($buffer);
             $request = $request->withBody($stream);
 
             return $stack($request);
-        }, function($error) {
+        }, function ($error) use ($stack, $request) {
             // request body of unknown size exceeding limit during buffering
             if ($error instanceof OverflowException) {
-                return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+                return $stack($request);
```
Comment (clue) [inline src/Middleware/RequestBodyBufferMiddleware.php]: Awesome, I think these changes make perfect sense :+1: 

One question that remains is, when should the next handler be called? It's my understanding that this middleware handler should still wait for the complete request to `end` before proceeding. This basically means that this middleware will wait for the complete request body anyway and that we simply ignore its contents.

Also, I wonder what should happen with `Content-Length` (remove?) and `getBody()->getSize()` (set to `0`?) etc. We should probably ensure to only pass updated values to the next consumer.

Can you update this to include some tests and documentation for this? :+1:

Comment (WyriHaximus) [reply src/Middleware/RequestBodyBufferMiddleware.php]: > One question that remains is, when should the next handler be called? It's my understanding that this middleware handler should still wait for the complete request to end before proceeding. This basically means that this middleware will wait for the complete request body anyway and that we simply ignore its contents.

Adding tests next so we can ensure the right behavior.

> Can you update this to include some tests and documentation for this? :+1:

It's on the list, but wanted code out first so we can discuss before spending time on documentation.

**Comment by WyriHaximus** [comment]
@clue added test that ensures the request isn't handled until the full request is in, also implemented that :tada: . Updating documentation next

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 2]
```diff
@@ -30,27 +30,35 @@ public function __construct($sizeLimit = null)
...
             if ($error instanceof OverflowException) {
-                return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+                return new Promise(function ($resolve, $reject) use ($stack, $request, $body) {
+                    $body->on('error', function ($error) use ($reject) {
+                        $reject($error);
+                    });
+                    $body->on('close', function () use ($stack, $request, $resolve) {
+                        $resolve($stack($request));
+                    });
+                });
```
Comment (clue) [inline src/Middleware/RequestBodyBufferMiddleware.php]: Changes LGTM, but this is currently missing a cancellation handler. Maybe use [`Stream\first()`](https://github.com/reactphp/promise-stream#first) instead?

Comment (WyriHaximus) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Good catch, I'll look into it :+1:

Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Not sure how relevant this here, what do you think about https://github.com/reactphp/promise-stream/pull/7?

**Comment by WyriHaximus** [comment]
@clue using `Stream\first() ` now and updated documentation

File: tests/Middleware/RequestBodyBufferMiddlewareTest.php [hunk 3]
```diff
@@ -79,17 +80,18 @@ public function testExcessiveSizeImmediatelyReturnsError413ForKnownSize()
...
-        );
+        ), $loop);
 
-        $this->assertSame(413, $response->getStatusCode());
+        $this->assertSame(200, $response->getStatusCode());
+        $this->assertSame('', $response->getBody()->getContents());
     }
 
-    public function testExcessiveSizeReturnsError413()
+    public function testExcessiveSizeBodyIsDisgardedTheRequestIsPassedDownToTheNextMiddleware()
```
Comment (jsor) [inline tests/Middleware/RequestBodyBufferMiddlewareTest.php]: Disgarded -> Dis**c**arded**And**

File: README.md [hunk 4]
```diff
@@ -694,8 +694,8 @@ configuration.
 configuration in most cases.)
 
 Any incoming request that has a request body that exceeds this limit will be
-rejected with a `413` (Request Entity Too Large) error message without calling
-the next middleware handlers.
+accepted but their request body will not be added to the request. Browsers consider 
+a request failed when a response is sent before the entire request has gone out.
```
Comment (clue) [inline README.md]: This is valid remark, but it is actually unrelated to this PR. I will remove this line for now and will make sure file a new ticket to keep track of this :+1:

**Comment by WyriHaximus** [comment]
@jsor ping!

File: README.md (hunk 4 above)
Comment (clue) [reply README.md]: See  #273 for follow-up discussion about this :+1:

**Pull request merged**